fpmsyncd/bgpmon.py

```diff
@@ -0,0 +1,171 @@
+#!/usr/bin/env python
+
+""""
+Description: bgpmon.py -- populating bgp related information in stateDB.
+    script is started by supervisord in bgp docker when the docker is started.
+
+    Initial creation of this daemon is to assist SNMP agent in obtaining the 
+    BGP related information for its MIB support. The MIB that this daemon is
+    assiting is for the CiscoBgp4MIB (Neighbor state only). If there are other
+    BGP related items that needs to be updated in a periodic manner in the 
+    future, then more can be added into this process.
+
+    The script check if there are any bgp activities by monitoring the bgp
+    frr.log file timestamp.  If activity is detected, then it will request bgp
+    neighbor state via vtysh cli interface. This bgp activity monitoring is 
+    done periodically (every 15 second). When triggered, it looks specifically
+    for the neighbor state in the json output of show ip bgp neighbors json
+    and update the state DB for each neighbor accordingly.
+    In order to not disturb and hold on to the State DB access too long and
+    removal of the stale neighbors (neighbors that was there previously on 
+    previous get request but no longer there in the current get request), a
+    "previous" neighbor dictionary will be kept and used to determine if there
+    is a need to perform update or the peer is stale to be removed from the
+    state DB
+"""
+
+import commands
+import json
+import os
+import syslog
+import swsssdk
+import time
+import traceback
+
+class BgpStateGet():
+    def __init__(self):
+        self.ipv4_n = []
+        self.ipv4_n_state = {}
+        self.ipv6_n = []
+        self.ipv6_n_state = {}
+        self.new_ipv4_n = []
+        self.new_ipv4_n_state = {}
+        self.new_ipv6_n = []
+        self.new_ipv6_n_state = {}
+        self.retrieve_fail_cnt = 0
+        self.cached_timestamp = 0
+        self.db = swsssdk.SonicV2Connector(host='127.0.0.1')
+        self.db.connect(self.db.STATE_DB, False)
+        self.db.delete_all_by_pattern(self.db.STATE_DB, "NEIGH_STATE_TABLE|*" )
+
+    # A quick way to check if there are anything happening within BGP is to
+    # check its log file has any activities. This is by checking its modified
+    # timestamp against the cached timestamp that we keep and if there is a
+    # difference, there is activity detected. In case the log file got wiped
+    # out, it will default back to constant pulling every 15 seconds
+    def bgp_activity_detected(self):
+        try:
+            timestamp = os.stat("/var/log/frr/frr.log").st_mtime
+            if timestamp != self.cached_timestamp:
+                #syslog.syslog("bgpmon timestamp diff detected: {}".format(timestamp))
+                self.cached_timestamp = timestamp
+                return True
+            else:
+                return False
+        except Exception, e:
```

Comment:
>Exception [](start = 15, length = 9)

Use more specific exception type

Reply:
@qiluo-msft Under normal operation I do not expect any exception resulting from attempting to read the timestamp of the frr.log file.  The only exception that may occurred is that someone accidentally renamed the log file or deleted the log file so the file path no longer exists.  Another possible exception perhaps has to do with access permission such as for some reason this daemon does not have the read permission and resulted to an exception.  Since the purpose is to "try" to use the FRR log file timestamp as a way to detect any BGP activities and if this mechanism temporarily not working, this daemon should still operate but just won't be as efficient as we desire it to be (became periodically reading out the bgp summary whether there is changes or not).  So perhaps it is ok not to be specific in this case as we don't really care about the specific reason why getting the frr log file timestamp failed.  We will just go on executing the BGP info pulling periodically anyway...  Let me know if you are ok with the above reasoning.

Reply:
Thanks for detailed analysis! I prefer conservative in order not to hide any issue. You extreme samples are the good ones for stop daemon and engage more investigation.

---
In reply to: [484672855](https://github.com/Azure/sonic-swss/pull/1429#discussion_r484672855) [](ancestors = 484672855)

Reply:
@qiluo-msft Agreed. I have changed to use more specific exception so it will not mask out important exceptions.